container-search/src/main/java/com/yahoo/vespa/streamingvisitors/VdsVisitor.java

```diff
@@ -102,10 +111,32 @@ public VisitorSession createVisitorSession(VisitorParameters params) throws Pars
         public LoadTypeSet getLoadTypeSet() {
             return loadTypes;
         }
+
+        /**
+         * Returns a single, shared instance of this class which is lazily created in a thread-safe
+         * manner the first time this method is invoked.
+         *
+         * May throw any config-related exception if subscription fails.
+         */
+        static MessageBusVisitorSessionFactory sharedInstance() {
+            var ref = instance.getAcquire();
```

Comment:
This extra check is not needed unless performance is really critical. A simpler approach would be to only keep the code inside the synchornized block and change `instance` to be a plain reference.

Reply:
Consider changing the members of `MessageBusVisitorSessionFactory` to be non-static, and instead enforce the singleton properties by having a single `MessageBusVisitorSessionFactory` instance as a static field in the outer class.

Reply:
If `MessageBusVisitorSessionFactory` is only used in context of JDisc, consider having it as a component instead. That is the idiomatic way of doing singletons in JDisc.

Reply:
I agree with your points and will consider them for a follow-up. For now I will use this TTAS approach since I try to avoid locks in serving paths if I can do so in a safe and contained manner.

I think what we really want is a simple "call once" library abstraction with similar semantics to C++11's function-level `static` or `call_once` functionality.